packages/stylist-macros/src/output/context_recorder.rs

```diff
@@ -14,13 +14,10 @@ enum AllowedUsage {
     // //       ------------- everything is static, do wrap style in Lazy
     // ```
     Static,
-    // TODO: we can probably avoid a few allocations if we track which parts
+    // We can avoid a few allocations if we track which parts
     // of the ast can be constructed statically (with const methods), which is
     // even stronger than constructing it in the global context in a Lazy.
-    // Should you decide to implement this, keep in mind to change Self::MAX
-    // and adjust the generation of cow-vec tokens. Also check the usages of
-    // MaybeStatic::statick if they can be upgraded to Const.
-    // Const,
+    Const,
```

Comment:
Do you mean `Static` and `Dynamic`? At least with the analysis of the superfluous `uses_static` mentioned above, the only place to call `uses_static` is sheet, which doesn't actually care, since there is no `const` option in sheet. Let me explain why conceptually three variants are needed:

- `Const` is ofc connected to `const` in rust, with very specific kinds of expressions that can be used in that context, so far so good. In case an expression that can not be in `const` expressions is used,
- `Static` would be everything that can be sensibly wrapped in a `static Lazy`. Note that I'm a bit unsure if this shouldn't actually be `const Lazy`, the difference is afaik in linkage. Anyway, the important point is that for an expression to qualify as `Static`, it can not mention any extra variables - which wouldn't be available in the closure given to `Lazy::new`.
- `Dynamic` does use local variables from the context and can not be sensibly wrapped in a `Lazy` because of that.

The fact that the expression analysis doesn't currently determine the `AllowedUsage` of an interpolation, apart from special casing string literals doesn't mean we should throw away the distinction between `Const` and `Static`. E.g. if the PR ends up not using `Bow`, then that would imply `Static` instead of `Const` for all parent ast nodes. Ultimately the sheet currently doesn't care, but the intermediate node do.

Reply:
`Ord` immensely clarifies the implementation of `uses_nested`, imo, with 3 variants that would otherwise be 6 or up to 9 cases?

Reply:
Follow up to `Ord`, not using the derive generates worse IR than what I would come up by hand, see e.g. https://godbolt.org/z/8KvGPPhsK Not that it matters in this case, I'll remove the usages of `Ord`.

Reply:
Maybe we can keep `Ord` but give the enum an explicit discriminator so it's less confusing?

Reply:
instead of `min`, the method is now called `joint_kind` and the usages of `<=` have been replaced by `matches!()`, the performance benefits of deriving `Ord` isn't worth it, imo.